### ISSUE TITLE
torch.median returns the smaller element when the median value lies between two elements.

### ISSUE THREAD

**Issue by heitorschueroff** [pytorch/pytorch#39520]
## 🐛 Bug

When the median value lies between two elements, torch.median returns the floor instead of the average.

## To Reproduce

1. start a python shell
2. run torch.tensor([1, 2, 3, 4]).median()
3. observe the result is 2 instead of 2.5

```
import torch
torch.tensor([1, 2, 3, 4]).median()
```

cc @jlin27 @mruberry

**Comment by ShawnZhong** [comment]
For your case, one of the reasons I can think of is that PyTorch uses `LongTensor` as the return type for `median` if the input type is `LongTensor`: 
```py
>>> torch.tensor([1, 2, 3, 4]).median().type()
'torch.LongTensor'
```

But even if you change the input type to `FloatType`, it's still hard to make it return the average since `median` can return the indices of the median value: 

```py
>>> torch.tensor([1., 2., 3., 4.]).median(dim=-1)
torch.return_types.median(
values=tensor(2.),
indices=tensor(1))
```

I think we'd better document this behavior to avoid confusion.

**Comment by colesbury** [comment]
1) At a minimum, we should document the current behavior
2) If no unique median exists, it would be good to use the typical convention (mean of the two middle elements). This would also match NumPy's behavior.

(2) is a little complicated given that we can return indices when `dim` is passed.

I think our current behavior fits the [formal definition](https://en.wikipedia.org/wiki/Median#Formal_definition) of a median, but it's probably not what most people would expect.

**Comment by ShawnZhong** [comment]
Aside from the indices, we might need some special handling for the gradient if we'd like to take the mean of the two middle elements.
